services/src/machine_learning/ml_tasks.rs

```diff
@@ -0,0 +1,202 @@
+    pub input_workflows: Vec<Workflow>,
+    pub label_workflow: Vec<Workflow>,
+    pub query: VectorQueryRectangle,
+}
+
+/// response of the machine learning model from workflow handler
+#[derive(Clone, Debug, Deserialize, Serialize, ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningModelFromWorkflowResult {
+    pub model_uuid: PathBuf,
+    pub model: Value,
+}
+
+impl TaskStatusInfo for MachineLearningModelFromWorkflowResult {}
+
+pub struct MachineLearningModelFromWorkflowTask<C: SessionContext> {
+    pub ctx: Arc<C>,
+    pub request: MLTrainRequest,
+    pub store_path: PathBuf,
+}
+
+impl<C: SessionContext> MachineLearningModelFromWorkflowTask<C> {
+    async fn process<A: Aggregatable<Data = Vec<f32>>>(
+        &self,
+    ) -> Result<MachineLearningModelFromWorkflowResult> {
+        let n_features = self.request.input_workflows.len();
+        let n_labels = self.request.label_workflow.len();
+
+        let mut feature_names = self.request.params.feature_names.clone();
+
+        // put the labels at the end of the feature name vec
+        feature_names.extend(Some(self.request.params.label_name.clone()));
+        let n_features_names = feature_names.len();
+
+        // make sure the correct number of names is provided
+        ensure!(
+            n_features >= 1 && n_labels >= 1 && (n_features + n_labels) == n_features_names,
+            super::xg_error::error::WrongNumberOfFeatureNamesProvided {
+                workflows: self.request.input_workflows.len(),
+                labels: self.request.label_workflow.len(),
+                names: feature_names.len()
+            }
+        );
+
+        let input_operators = get_operators_from_workflows(
+            &self.request.input_workflows,
+            &self.request.label_workflow,
+        )?;
+
+        let mut exe_ctx = self.ctx.execution_context()?;
+
+        let typed_query_processors = get_query_processors(input_operators, &exe_ctx).await?;
+
+        let query = self.request.query;
+        let query_ctx = self.ctx.query_context()?;
+
+        let n_rasters = typed_query_processors.len();
+
+        let mut aggregator_vec: Vec<_> = Vec::with_capacity(n_rasters);
+
+        (0..n_rasters).into_iter().for_each(|_| {
+            let data = Vec::new();
+            let agg: A = Aggregatable::new(data, self.request.params.memory_limit);
+            aggregator_vec.push(agg);
+        });
+
+        let mut feature_data: Vec<MachineLearningFeature> = accumulate_raster_data(
+            feature_names,
+            typed_query_processors,
+            query,
+            &query_ctx,
+            &mut aggregator_vec,
+        )
+        .await?;
+
+        // extract the label data from the feature data vec
+        let mut label_data = feature_data.drain(n_features..).collect();
+
+        //TODO: this could be wrapped by a match or even a dedicated ml model struct,
+        // which unifies the different kind of models that could be generated
```

Comment:
maybe not "unifies" but "delegates based on the kind of model to a different task"?

Reply:
commit msg: chore: rephrase TODO comment on ml request delegation
ee5c755c